src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepDynamicContext.java

```diff
@@ -90,12 +91,33 @@ void resume(StepContext context) throws Exception {
         if (executor != null) {
             throw new IllegalStateException("Already resumed");
         }
-        Queue.Item item = Queue.getInstance().schedule2(task, 0).getItem();
-        if (item == null) {
-            // TODO should also report when !ScheduleResult.created, since that is arguably an error
-            throw new IllegalStateException("queue refused " + task);
+        // If Jenkins restarts or crashes while we are waiting below, on the next startup the task may already be in
+        // the queue, in which case we should reuse it instead of scheduling a second task.
+        AtomicReference<Queue.Item> itemRef = new AtomicReference<>();
+        Queue.withLock(() -> {
+            for (Queue.Item item : Queue.getInstance().getItems()) {
+                if (item.task instanceof ExecutorStepExecution.PlaceholderTask) {
+                    ExecutorStepExecution.PlaceholderTask itemTask = (ExecutorStepExecution.PlaceholderTask) item.task;
+                    if (task.equals(itemTask)) {
+                        itemRef.set(item);
+                        break;
+                    }
+                }
+            }
+        });
+        // TODO: Do we need to check executor slots too in case there was a task in the queue but it has already started?
+        if (itemRef.get() == null) {
+            Queue.Item item = Queue.getInstance().schedule2(task, 0).getItem();
+            if (item == null) {
+                // TODO should also report when !ScheduleResult.created, since that is arguably an error
+                throw new IllegalStateException("queue refused " + task);
+            }
+            LOGGER.fine(() -> "scheduled " + item + " for " + path + " on " + node);
+            itemRef.set(item);
+        } else {
+            LOGGER.fine(() -> "reusing " + itemRef.get() + ", which was already in the queue for " + path + " on " + node);
         }
-        LOGGER.fine(() -> "scheduled " + item + " for " + path + " on " + node);
+        Queue.Item item = itemRef.get();
```

Comment:
BTW it might be more legible to replace the `AtomicReference` idiom with a `Stream` usage, so you do not need two variables.

Reply:
[456010d](https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/344/commits/456010d9c970e0e2ee516d5f07b63dd5425f45ca) removes the need for the reference.